File: services/payment-service/src/api/validation.ts
import { config } from '../config';
import type { AccountLimits } from '../domain/limits';
import { isValidAmount, type SimulateMode } from '../domain/payment';
import { badRequest } from '../lib/http-error';

/**
 * The trust boundary. Nothing past this file assumes anything about the shape
 * of a request - if a value got here, it is the right type and within range.
 */

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export const isUuid = (value: unknown): value is string =>
  typeof value === 'string' && UUID_RE.test(value);

export function requireUuid(value: unknown, code: string): string {
  if (!isUuid(value)) throw badRequest(code);
  return value;
}

/**
 * Clamped both ways: a negative limit would become a negative range index,
 * which some stores read as "from the end" and quietly return the wrong slice.
 */
export const clampLimit = (raw: unknown, fallback: number, max: number): number =>
  Math.min(Math.max(Number(raw ?? fallback) || fallback, 1), max);

export function parseAccountName(value: unknown): string {
  if (typeof value !== 'string' || value.trim() === '') throw badRequest('NAME_REQUIRED');
  const name = value.trim();
  if (name.length > config.limits.nameLength) throw badRequest('NAME_TOO_LONG');
  return name;
}

export function parseOpeningBalance(value: unknown): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) {
    throw badRequest('INVALID_INITIAL_BALANCE');
  }
  return value as number;
}

/** A limit is a non-negative whole number of cents; zero means "blocked". */
function parseCap(value: unknown, code: string, max: number): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0 || (value as number) > max) {
    throw badRequest(code);
  }
  return value as number;
}

export function parseLimits(body: Record<string, unknown>): AccountLimits {
  return {
    maxPaymentCents: parseCap(
      body.maxPaymentCents,
      'INVALID_MAX_PAYMENT',
      config.limits.maxLimitCents,
    ),
    dailyLimitCents: parseCap(
      body.dailyLimitCents,
      'INVALID_DAILY_LIMIT',
      config.limits.maxLimitCents,
    ),
    velocityMax: parseCap(body.velocityMax, 'INVALID_VELOCITY_MAX', 10_000),
  };
}

export function parseNote(value: unknown): string | null {
  if (value === undefined || value === null) return null;
  if (typeof value !== 'string') throw badRequest('INVALID_NOTE');
  if (value.length > config.limits.noteLength) throw badRequest('NOTE_TOO_LONG');
  return value.trim() || null;
}

/**
 * How the settle leg should be made to fail, for demonstrating the saga.
 * `simulate` is the current field; the older boolean `simulateFailure: true`
 * is still accepted and means "permanent".
 */
export function parseSimulateMode(
  simulate: unknown,
  simulateFailure: unknown,
): SimulateMode {
  let mode: SimulateMode = 'NONE';

  if (simulateFailure !== undefined) {
    if (typeof simulateFailure !== 'boolean') throw badRequest('INVALID_SIMULATE_FAILURE');
    if (simulateFailure) mode = 'PERMANENT';
  }

  if (simulate !== undefined) {
    const wanted = String(simulate).toUpperCase();
    if (!['NONE', 'TRANSIENT', 'PERMANENT'].includes(wanted)) {
      throw badRequest('INVALID_SIMULATE_MODE');
    }
    mode = wanted as SimulateMode;
  }

  return mode;
}

export function parseIdempotencyKey(header: string | undefined): string | null {
  const key = header?.trim() || null;
  if (key && key.length > config.limits.idempotencyKeyLength) {
    throw badRequest('IDEMPOTENCY_KEY_TOO_LONG');
  }
  return key;
}

export interface ValidatedTransfer {
  fromAccountId: string;
  toAccountId: string;
  amountCents: number;
}

export function parseTransfer(body: Record<string, unknown>): ValidatedTransfer {
  const fromAccountId = String(body.fromAccountId);
  const toAccountId = String(body.toAccountId);

  if (!isUuid(fromAccountId) || !isUuid(toAccountId)) throw badRequest('INVALID_ACCOUNT_ID');
  if (!isValidAmount(body.amountCents)) throw badRequest('INVALID_AMOUNT');
  if (fromAccountId === toAccountId) throw badRequest('SAME_ACCOUNT');

  // The clearing and funding accounts are plumbing. Letting a client move
  // money in or out of them directly would put the ledger's invariants at the
  // mercy of the API.
  const systemIds: string[] = [
    config.systemAccounts.clearingId,
    config.systemAccounts.fundingId,
  ];
  if (systemIds.includes(fromAccountId) || systemIds.includes(toAccountId)) {
    throw badRequest('SYSTEM_ACCOUNT_NOT_PAYABLE');
  }

  return { fromAccountId, toAccountId, amountCents: body.amountCents };
}

File: services/payment-service/src/main/java/com/ledgerflow/payment/api/Validation.java
package com.ledgerflow.payment.api;

import com.ledgerflow.payment.config.Config;
import com.ledgerflow.payment.domain.Limits.AccountLimits;
import com.ledgerflow.payment.domain.Payments;
import com.ledgerflow.payment.domain.SimulateMode;
import com.ledgerflow.payment.lib.HttpError;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * The trust boundary. Nothing past this class assumes anything about the shape
 * of a request - if a value got here, it is the right type and within range.
 */
public final class Validation {

  private Validation() {}

  private static final Pattern UUID_RE =
      Pattern.compile(
          "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
          Pattern.CASE_INSENSITIVE);

  public static boolean isUuid(String value) {
    return value != null && UUID_RE.matcher(value).matches();
  }

  public static String requireUuid(String value, String code) {
    if (!isUuid(value)) throw HttpError.badRequest(code);
    return value;
  }

  /**
   * Clamped both ways: a negative limit would become a negative range index,
   * which some stores read as "from the end" and quietly return the wrong slice.
   */
  public static int clampLimit(String raw, int fallback, int max) {
    double wanted = fallback;
    if (raw != null) {
      try {
        double parsed = Double.parseDouble(raw.trim());
        // A zero or unparseable limit means "unspecified", as Number(x) || fallback did.
        wanted = Double.isFinite(parsed) && parsed != 0 ? parsed : fallback;
      } catch (NumberFormatException e) {
        wanted = fallback;
      }
    }
    return (int) Math.min(Math.max(wanted, 1), max);
  }

  public static String parseAccountName(Object value) {
    if (!(value instanceof String text) || text.trim().isEmpty()) {
      throw HttpError.badRequest("NAME_REQUIRED");
    }
    String name = text.trim();
    if (name.length() > Config.Limits.NAME_LENGTH) throw HttpError.badRequest("NAME_TOO_LONG");
    return name;
  }

  public static long parseOpeningBalance(Object value) {
    Long balance = Payments.asSafeInteger(value);
    if (balance == null || balance < 0) throw HttpError.badRequest("INVALID_INITIAL_BALANCE");
    return balance;
  }

  /** A limit is a non-negative whole number of cents; zero means "blocked". */
  private static long parseCap(Object value, String code, long max) {
    Long cap = Payments.asSafeInteger(value);
    if (cap == null || cap < 0 || cap > max) throw HttpError.badRequest(code);
    return cap;
  }

  public static AccountLimits parseLimits(Map<String, Object> body) {
    return new AccountLimits(
        parseCap(body.get("maxPaymentCents"), "INVALID_MAX_PAYMENT", Config.Limits.MAX_LIMIT_CENTS),
        parseCap(body.get("dailyLimitCents"), "INVALID_DAILY_LIMIT", Config.Limits.MAX_LIMIT_CENTS),
        (int) parseCap(body.get("velocityMax"), "INVALID_VELOCITY_MAX", 10_000));
  }

  public static String parseNote(Object value) {
    if (value == null) return null;
    if (!(value instanceof String text)) throw HttpError.badRequest("INVALID_NOTE");
    if (text.length() > Config.Limits.NOTE_LENGTH) throw HttpError.badRequest("NOTE_TOO_LONG");
    String trimmed = text.trim();
    return trimmed.isEmpty() ? null : trimmed;
  }

  /**
   * How the settle leg should be made to fail, for demonstrating the saga.
   * `simulate` is the current field; the older boolean `simulateFailure: true`
   * is still accepted and means "permanent".
   */
  public static SimulateMode parseSimulateMode(Object simulate, Object simulateFailure) {
    SimulateMode mode = SimulateMode.NONE;

    if (simulateFailure != null) {
      if (!(simulateFailure instanceof Boolean flag)) {
        throw HttpError.badRequest("INVALID_SIMULATE_FAILURE");
      }
      if (flag) mode = SimulateMode.PERMANENT;
    }

    if (simulate != null) {
      String wanted = String.valueOf(simulate).toUpperCase();
      try {
        mode = SimulateMode.valueOf(wanted);
      } catch (IllegalArgumentException e) {
        throw HttpError.badRequest("INVALID_SIMULATE_MODE");
      }
    }

    return mode;
  }

  public static String parseIdempotencyKey(String header) {
    String key = header == null || header.trim().isEmpty() ? null : header.trim();
    if (key != null && key.length() > Config.Limits.IDEMPOTENCY_KEY_LENGTH) {
      throw HttpError.badRequest("IDEMPOTENCY_KEY_TOO_LONG");
    }
    return key;
  }

  public record ValidatedTransfer(String fromAccountId, String toAccountId, long amountCents) {}

  public static ValidatedTransfer parseTransfer(Map<String, Object> body) {
    String fromAccountId = String.valueOf(body.get("fromAccountId"));
    String toAccountId = String.valueOf(body.get("toAccountId"));

    if (!isUuid(fromAccountId) || !isUuid(toAccountId)) {
      throw HttpError.badRequest("INVALID_ACCOUNT_ID");
    }
    Long amountCents = Payments.asSafeInteger(body.get("amountCents"));
    if (amountCents == null || !Payments.isValidAmount(amountCents.longValue())) {
      throw HttpError.badRequest("INVALID_AMOUNT");
    }
    if (fromAccountId.equals(toAccountId)) throw HttpError.badRequest("SAME_ACCOUNT");

    // The clearing and funding accounts are plumbing. Letting a client move
    // money in or out of them directly would put the ledger's invariants at the
    // mercy of the API.
    List<String> systemIds =
        List.of(Config.SystemAccounts.CLEARING_ID, Config.SystemAccounts.FUNDING_ID);
    if (systemIds.contains(fromAccountId) || systemIds.contains(toAccountId)) {
      throw HttpError.badRequest("SYSTEM_ACCOUNT_NOT_PAYABLE");
    }

    return new ValidatedTransfer(fromAccountId, toAccountId, amountCents);
  }
}
